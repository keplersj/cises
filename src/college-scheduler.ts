import { getProfessorFromSchool } from "./lib/rmp";

const schoolName = "University of Utah";

async function getProfessorFromScheduleClassModal() {
  const elements = document.querySelectorAll(
    "html body.static-header.modal-open div#base-modal.modal.fade.center-lg.in div.modal-dialog.modal-lg div.modal-content div.modal-body div.row div.col-md-6 ul.section-details li.persist"
  );

  for (let element of elements) {
    if ((element as HTMLElement).innerText.match(/Instructor: /)) {
      const name = ((element as HTMLElement).lastChild as Text).data;
      return await getProfessorFromSchool(name, schoolName);
    }
  }

  return undefined;
}

const bodyObserver = new MutationObserver(mutationsList => {
  for (var mutation of mutationsList) {
    if (mutation.type == "childList") {
      mutation.addedNodes.forEach(node => {
        if ((node as Element).id === "base-modal") {
          getProfessorFromScheduleClassModal().then(console.log);
        }
      });
    }
  }
});

// Start observing the target node for configured mutations
bodyObserver.observe(document.body, {
  attributes: false,
  childList: true,
  subtree: false
});
