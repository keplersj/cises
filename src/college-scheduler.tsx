import React from "react";
import ReactDOM from "react-dom";
import { getProfessorFromSchool } from "./lib/rmp";
import { ScheduleBuilderRMP } from "./views/schedule_builder_rmp";

const schoolName = "University of Utah";

async function getProfessorFromScheduleClassModal() {
  const elements = document.querySelectorAll(
    "html body.static-header.modal-open div#base-modal.modal.fade.center-lg.in div.modal-dialog.modal-lg div.modal-content div.modal-body div.row div.col-md-6 ul.section-details li.persist"
  );

  for (let element of elements) {
    if ((element as HTMLLIElement).innerHTML.match(/Instructor:/)) {
      const name = ((element as HTMLLIElement).lastChild as Text).data;
      return await getProfessorFromSchool(name, schoolName);
    }
  }

  return undefined;
}

async function renderProfInfoInModal() {
  const professor = await getProfessorFromScheduleClassModal();

  if (professor) {
    const container = document.querySelector(
      ".modal-body > div:nth-child(1) > div:nth-child(2)"
    );
    const rmpComponentContainer = document.createElement("div");
    ReactDOM.render(
      <ScheduleBuilderRMP
        rmpId={professor.pk_id}
        rating={professor.averageratingscore_rf}
        tags={professor.tag_s_mv}
      />,
      rmpComponentContainer
    );
    container && container.appendChild(rmpComponentContainer);
  }
}

const bodyObserver = new MutationObserver(mutationsList => {
  mutationsList.forEach(mutation => {
    Array.from(mutation.addedNodes.values())
      .filter(node => (node as Element).id === "base-modal")
      .forEach(renderProfInfoInModal);
  });
});

// Start observing the target node for configured mutations
bodyObserver.observe(document.body, {
  childList: true
});
