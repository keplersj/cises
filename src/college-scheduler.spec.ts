import "./college-scheduler";

describe("College Scheduler Content Script", () => {
  beforeEach(() => {
    fetchMock.resetMocks();
    document.body.innerHTML = "";
    document.body.className = "";
  });

  it("injects our component into the document when the `#base-modal` element is created", cb => {
    expect(document.body).toMatchSnapshot();

    fetchMock.mockResponseOnce(
      `noCB({ response: { docs: [${JSON.stringify({
        city_state_s: "",
        averageratingscore_rf: 2.0,
        pk_id: 2.0,
        schoolcity_s: "",
        schoolstate_full_s: "",
        pict_thumb_name_s: "",
        id: "",
        pageviews_i: 2.0,
        averagehelpfulscore_rf: 2.0,
        schoolcountry_s: "",
        schoolname_s: "",
        status_i: 2.0,
        teachermiddlename_t: "",
        averagehotscore_rf: 2.0,
        schoolstate_s: "",
        rated_date_dt: "",
        teacherfullname_s: "",
        teacherdepartment_s: "",
        total_number_of_ratings_i: 2.0,
        visible_i: 2.0,
        content_type_s: "",
        averageeasyscore_rf: 2.0,
        schoolid_s: "",
        teacherfirstname_t: "",
        teacherlastname_t: "",
        averageclarityscore_rf: 2.0,
        schoolwebpage_s: "",
        timestamp: "",
        tag_s_mv: ["", "", ""],
        tag_id_s_mv: ["", "", ""]
      })}] } })`
    );

    const baseModal = document.createElement("div");
    baseModal.id = "base-modal";
    baseModal.className = "modal fade center-lg in";
    baseModal.innerHTML = `
        <div class="modal-dialog modal-lg">
            <div class="modal-content">
                <div class="modal-body">
                    <div class="row">
                        <div class="col-md-6">
                            <ul class="section-details">
                                <li class="persist">
                                    <strong>Instructor:</strong> Name McNamey
                                </li>
                            </ul>
                        </div>
                        <div class="col-md-6">
                            
                        </div>
                    </div>
                </div>
            </div>
        </div>
    `;
    document.body.className = "static-header modal-open";
    document.body.appendChild(baseModal);

    setTimeout(() => {
      expect(document.body).toMatchSnapshot();
      expect(document.querySelector("#cises-rmp")).toBeTruthy();
      cb();
    });
  });

  it("does not inject our component if Schedule Builder doesn't include an instructor in `#base-modal`", cb => {
    expect(document.body).toMatchSnapshot();

    const baseModal = document.createElement("div");
    baseModal.id = "base-modal";
    baseModal.className = "modal fade center-lg in";
    baseModal.innerHTML = `
        <div class="modal-dialog modal-lg">
            <div class="modal-content">
                <div class="modal-body">
                    <div class="row">
                        <div class="col-md-6">
                            <ul class="section-details">
                                <li class="persist">
                                    <strong>Class Section:</strong> FAKE 1234
                                </li>
                            </ul>
                        </div>
                        <div class="col-md-6">
                            
                        </div>
                    </div>
                </div>
            </div>
        </div>
    `;
    document.body.className = "static-header modal-open";
    document.body.appendChild(baseModal);

    setTimeout(() => {
      expect(document.body).toMatchSnapshot();
      expect(document.querySelector("#cises-rmp")).toBeNull();
      cb();
    });
  });
});
