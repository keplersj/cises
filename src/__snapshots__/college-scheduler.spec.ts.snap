// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`College Scheduler Content Script does not inject our component if Schedule Builder doesn't include an instructor in \`#base-modal\` 1`] = `
<body
  class=""
/>
`;

exports[`College Scheduler Content Script does not inject our component if Schedule Builder doesn't include an instructor in \`#base-modal\` 2`] = `
<body
  class="static-header modal-open"
>
  <div
    class="modal fade center-lg in"
    id="base-modal"
  >
    
        
    <div
      class="modal-dialog modal-lg"
    >
      
            
      <div
        class="modal-content"
      >
        
                
        <div
          class="modal-body"
        >
          
                    
          <div
            class="row"
          >
            
                        
            <div
              class="col-md-6"
            >
              
                            
              <ul
                class="section-details"
              >
                
                                
                <li
                  class="persist"
                >
                  
                                    
                  <strong>
                    Class Section:
                  </strong>
                   FAKE 1234
                                
                </li>
                
                            
              </ul>
              
                        
            </div>
            
                        
            <div
              class="col-md-6"
            >
              
                            
                        
            </div>
            
                    
          </div>
          
                
        </div>
        
            
      </div>
      
        
    </div>
    
    
  </div>
</body>
`;

exports[`College Scheduler Content Script injects our component into the document when the \`#base-modal\` element is created 1`] = `
<body
  class=""
/>
`;

exports[`College Scheduler Content Script injects our component into the document when the \`#base-modal\` element is created 2`] = `
.emotion-5 {
  border-radius: 15px;
  padding: 10px;
  background-color: #c0c0c0;
  color: #ffffff;
}

.emotion-0 {
  font-weight: 900;
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0:hover {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-0:active {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-1 {
  background-color: #a0a0a0;
  margin: 2px;
  padding-left: 1px;
  padding-right: 1px;
}

<body
  class="static-header modal-open"
>
  <div
    class="modal fade center-lg in"
    id="base-modal"
  >
    
        
    <div
      class="modal-dialog modal-lg"
    >
      
            
      <div
        class="modal-content"
      >
        
                
        <div
          class="modal-body"
        >
          
                    
          <div
            class="row"
          >
            
                        
            <div
              class="col-md-6"
            >
              
                            
              <ul
                class="section-details"
              >
                
                                
                <li
                  class="persist"
                >
                  
                                    
                  <strong>
                    Instructor:
                  </strong>
                   Name McNamey
                                
                </li>
                
                            
              </ul>
              
                        
            </div>
            
                        
            <div
              class="col-md-6"
            >
              
                            
                        
              <div>
                <div
                  class="emotion-5"
                  id="cises-rmp"
                >
                  <a
                    class="emotion-0"
                    href="https://www.ratemyprofessors.com/ShowRatings.jsp?tid=2"
                  >
                    Info from Rate My Professors:
                  </a>
                  <br />
                  <strong>
                    Average Rating: 
                  </strong>
                  <span>
                    2
                  </span>
                  <br />
                  <strong>
                    Tags:
                  </strong>
                  <br />
                  <div
                    class="emotion-4"
                  >
                    <span
                      class="emotion-1"
                    />
                    <span
                      class="emotion-1"
                    />
                    <span
                      class="emotion-1"
                    />
                  </div>
                </div>
              </div>
            </div>
            
                    
          </div>
          
                
        </div>
        
            
      </div>
      
        
    </div>
    
    
  </div>
</body>
`;
